Assets/MRTK/Services/InputSystem/MixedRealityInputSystem.cs

```diff
@@ -575,11 +575,60 @@ private void HandleFocusEvent(GameObject eventTarget, FocusEventData focusEventD
             // If event was not handled by modal, pass it on to the current focused object
             if (focusedObject != null)
             {
-                ExecuteEvents.ExecuteHierarchy(focusedObject, baseInputEventData, eventHandler);
+                // Handling propagation
+                var eventPropagationData = baseInputEventData as IMixedRealityEventPropagationData;
+                if (eventPropagationData != null)
+                {
+                    PropagateEvent<T>(focusedObject, baseInputEventData, eventHandler);
+                }
+                // Events with no propagation configuration
+                else
+                {
+                    ExecuteEvents.ExecuteHierarchy(focusedObject, baseInputEventData, eventHandler);
+                }
             }
             return modalEventHandled;
         }
 
+        /// <summary>
+        /// Propagates an event down and up the target object hierarchy if the event supports propagation.
+        /// Propagation is not halting in case of event data already used. Handlers are expected to check use of event.
+        /// </summary>
+        private void PropagateEvent<T>(GameObject targetObject, BaseInputEventData baseInputEventData, ExecuteEvents.EventFunction<T> eventHandler) where T : IEventSystemHandler
+        {
+            if (!(baseInputEventData is IMixedRealityEventPropagationData eventPropagationData))
+            {
+                return;
+            }
+
+            var targetHierarchy = targetObject.GetComponentsInParent<Transform>();
+
+            // Trickles Down
+            if (eventPropagationData.Propagation.HasFlag(EventPropagation.TricklesDown))
+            {
+                for (int i = targetHierarchy.Length - 1; i > 0; i--)
+                {
+                    eventPropagationData.Phase = PropagationPhase.TrickeDown;
+                    ExecuteEvents.Execute(targetHierarchy[i].gameObject, baseInputEventData, eventHandler);
```

Comment:
I wonder if it's worth adding a check to `ExecuteEvents.CanHandleEvent<T>(GameObject)` before executing the event? Not sure if the perf ends up better or worse, but it could be worth checking,

Reply:
Approach was changed.